src/constraint_aware_append.c

```diff
@@ -0,0 +1,384 @@
+	CustomScan *cscan = (CustomScan *) node->ss.ps.plan;
+	Plan	   *subplan = copyObject(state->subplan);
+	List	   *append_rel_info = lsecond(cscan->custom_private);
+	List	   *restrictinfos = constify_restrictinfos(lthird(cscan->custom_private));
+	List	  **appendplans,
+			   *old_appendplans;
+	ListCell   *lc_plan,
+			   *lc_info;
+
+	switch (nodeTag(subplan))
+	{
+		case T_Append:
+			{
+				Append	   *append = (Append *) subplan;
+
+				old_appendplans = append->appendplans;
+				append->appendplans = NIL;
+				appendplans = &append->appendplans;
+				break;
+			}
+		case T_MergeAppend:
+			{
+				MergeAppend *append = (MergeAppend *) subplan;
+
+				old_appendplans = append->mergeplans;
+				append->mergeplans = NIL;
+				appendplans = &append->mergeplans;
+				break;
+			}
+		default:
+			elog(ERROR, "Invalid plan %d", nodeTag(subplan));
+	}
+
+	forboth(lc_plan, old_appendplans, lc_info, append_rel_info)
+	{
+		Scan	   *scan = lfirst(lc_plan);
+		AppendRelInfo *appinfo = lfirst(lc_info);
+		RangeTblEntry *rte = rt_fetch(scan->scanrelid, estate->es_range_table);
+
+		switch (nodeTag(scan))
+		{
+			case T_SeqScan:
+			case T_SampleScan:
+			case T_IndexScan:
+			case T_IndexOnlyScan:
+			case T_BitmapIndexScan:
+			case T_BitmapHeapScan:
+			case T_TidScan:
+			case T_SubqueryScan:
+			case T_FunctionScan:
+			case T_ValuesScan:
+			case T_CteScan:
+			case T_WorkTableScan:
+			case T_ForeignScan:
+			case T_CustomScan:
+
+				/*
+				 * If this is a base rel (chunk), check if it can be excluded
+				 * from the scan. Otherwise, fall through.
+				 */
+				if (rte->rtekind == RTE_RELATION &&
+					rte->relkind == RELKIND_RELATION &&
+					!rte->inh &&
+					excluded_by_constraint(rte, appinfo, restrictinfos))
+					break;
+			default:
+				*appendplans = lappend(*appendplans, scan);
+		}
+	}
+
+	node->custom_ps = list_make1(ExecInitNode(subplan, estate, eflags));
+}
+
+static TupleTableSlot *
+ca_append_exec(CustomScanState *node)
+{
+	TupleTableSlot *subslot;
+	TupleTableSlot *resultslot;
+	ExprContext *econtext = node->ss.ps.ps_ExprContext;
+	ExprDoneCond isDone;
+
+	if (node->ss.ps.ps_TupFromTlist)
+	{
+		resultslot = ExecProject(node->ss.ps.ps_ProjInfo, &isDone);
+
+		if (isDone == ExprMultipleResult)
+			return resultslot;
+
+		node->ss.ps.ps_TupFromTlist = false;
+	}
+
+	ResetExprContext(econtext);
+
+	while (true)
+	{
+		subslot = ExecProcNode(linitial(node->custom_ps));
+
+		if (TupIsNull(subslot))
+			return NULL;
+
+		if (!node->ss.ps.ps_ProjInfo)
+			return subslot;
+
+		econtext->ecxt_scantuple = subslot;
+
+		resultslot = ExecProject(node->ss.ps.ps_ProjInfo, &isDone);
+
+		if (isDone != ExprEndResult)
+		{
+			node->ss.ps.ps_TupFromTlist = (isDone == ExprMultipleResult);
+			return resultslot;
+		}
+	}
+}
+
+static void
+ca_append_end(CustomScanState *node)
+{
+	ExecEndNode(linitial(node->custom_ps));
+}
+
+static void
+ca_append_rescan(CustomScanState *node)
```

Comment:
I don't quite know how this works. When does a rescan happen? Is it possible to trigger this to happen in tests?